stan/math/rev/functor/algebra_solver_powell.hpp

```diff
@@ -21,57 +22,77 @@ namespace math {
+  const Eigen::Matrix<T, Eigen::Dynamic, 1> y_val_;
+  /** System functor (f_) w.r.t. inputs (y_) */
+  Fy fy_;
+  /** array of parameters */
+  vari** y_;
   /** number of unknowns */
-  int x_size_;
-  /** vector of solution */
-  vari** theta_;
-  /** Jacobian of the solution w.r.t parameters */
-  double* Jx_y_;
+  const int x_size_;
+  /** value of unknowns */
+  const Eigen::VectorXd& x_val_;
+  /** Hybrj functor solver (f_) w.r.t. outputs (x_) */
+  Fx fx_;
+  /** array of unknowns */
+  vari** x_;
+  /** Jacobian of f w.r.t. outputs (x_) */
+  const Eigen::MatrixXd Jf_x_;
 
-  algebra_solver_vari(const Fs& fs, const F& f, const Eigen::VectorXd& x,
-                      const Eigen::Matrix<T, Eigen::Dynamic, 1>& y,
-                      const std::vector<double>& dat,
-                      const std::vector<int>& dat_int,
-                      const Eigen::VectorXd& theta_dbl, Fx& fx,
-                      std::ostream* msgs)
-      : vari(theta_dbl(0)),
-        y_(ChainableStack::instance_->memalloc_.alloc_array<vari*>(y.size())),
+  algebra_solver_vari(Fy& fy, const Eigen::Matrix<T, Eigen::Dynamic, 1>& y,
+                      Fx& fx, const Eigen::VectorXd& x)
+      : vari(x(0)),
         y_size_(y.size()),
+        y_val_(y),
+        fy_(fy),
+        y_(ChainableStack::instance_->memalloc_.alloc_array<vari*>(y_size_)),
         x_size_(x.size()),
-        theta_(
-            ChainableStack::instance_->memalloc_.alloc_array<vari*>(x_size_)),
-        Jx_y_(ChainableStack::instance_->memalloc_.alloc_array<double>(
-            x_size_ * y_size_)) {
+        x_val_(x),
+        Jf_x_(fx.get_jacobian(x_val_)),
+        fx_(fx),
+        x_(ChainableStack::instance_->memalloc_.alloc_array<vari*>(x_size_)) {
     using Eigen::Map;
     using Eigen::MatrixXd;
-    for (int i = 0; i < y.size(); ++i) {
+
+    for (int i = 0; i < y_size_; ++i) {
       y_[i] = y(i).vi_;
     }
 
-    theta_[0] = this;
-    for (int i = 1; i < x.size(); ++i) {
-      theta_[i] = new vari(theta_dbl(i), false);
+    x_[0] = this;
+    for (int i = 1; i < x_size_; ++i) {
+      x_[i] = new vari(x(i), false);
     }
-
-    // Compute the Jacobian and store in array, using the
-    // implicit function theorem, i.e. Jx_y = Jf_y / Jf_x
-    using f_y = hybrj_functor_solver<Fs, F, double, double>;
-    Map<MatrixXd>(&Jx_y_[0], x_size_, y_size_)
-        = -mdivide_left(fx.get_jacobian(theta_dbl),
-                        f_y(fs, f, theta_dbl, value_of(y), dat, dat_int, msgs)
-                            .get_jacobian(value_of(y)));
   }
 
   void chain() {
-    for (int j = 0; j < y_size_; j++) {
-      for (int i = 0; i < x_size_; i++) {
-        y_[j]->adj_ += theta_[i]->adj_ * Jx_y_[j * x_size_ + i];
+    using Eigen::Dynamic;
+    using Eigen::Matrix;
+    using Eigen::MatrixXd;
+    using Eigen::VectorXd;
+
+    // Compute (transpose of) specificities with respect to x.
+    VectorXd x_bar_(x_size_);  // TODO: Is this zeroing out memory?
```

Comment:
No, to initialize to zero do:

```
VectorXd x_bar = VectorXd::Zero(x_size_);
```

But you don't need to initialize it since you're writing in the adjoints from `x_` here.

Reply:
Right, that makes sense! I was just wanted to check that I *wasn't* zeroing out the memory since it all gets immediately overwritten anyway.